src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -0,0 +1,448 @@
+        {
+            switch (ch)
+            {
+                case '+': case '-':
+                case 'i': case 'I':
+                case 'm': case 'M':
+                case 'n': case 'N':
+                case 's': case 'S':
+                case 'x': case 'X':
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        public RegexToken ScanHexCharacters(int count)
+        {
+            var start = Position;
+            var beforeSlash = start - 2;
+
+            // Make sure we're right after the \x or \u.
+            Debug.Assert(Text[beforeSlash].Char == '\\');
+            Debug.Assert(Text[beforeSlash + 1].Char == 'x' || Text[beforeSlash + 1].Char == 'u');
+
+            for (int i = 0; i < count; i++)
+            {
+                if (Position < Text.Length && IsHexChar(this.CurrentChar))
+                {
+                    Position++;
+                }
+            }
+
+            var result = CreateToken(
+                RegexKind.TextToken, ImmutableArray<RegexTrivia>.Empty, GetSubPattern(start, Position));
+
+            var length = Position - start;
+            if (length != count)
+            {
+                result = result.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Insufficient_hexadecimal_digits,
+                    TextSpan.FromBounds(Text[beforeSlash].Span.Start, Text[Position - 1].Span.End)));
+            }
+
+            return result;
+        }
+
+        private static bool IsHexChar(char ch)
+            => IsDecimalDigit(ch) ||
+               (ch >= 'a' && ch <= 'f') ||
+               (ch >= 'A' && ch <= 'F');
+
+        private static bool IsDecimalDigit(char ch)
+            => ch >= '0' && ch <= '9';
+
+        private static bool IsOctalDigit(char ch)
+            => ch >= '0' && ch <= '7';
+
+        public RegexToken ScanOctalCharacters(RegexOptions options)
+        {
+            var start = Position;
+            var beforeSlash = start - 1;
+
+            // Make sure we're right after the \ or \.
+            Debug.Assert(Text[beforeSlash].Char == '\\');
+
+            const int maxChars = 3;
+            int currentVal = 0;
+
+            for (int i = 0; i < maxChars; i++)
+            {
+                if (Position < Text.Length && IsOctalDigit(this.CurrentChar))
+                {
+                    var octalVal = this.CurrentChar - '0';
+                    Debug.Assert(octalVal >= 0 && octalVal <= 7);
+                    currentVal *= 8;
+                    currentVal += octalVal;
+
+                    Position++;
+
+                    // Ecmascript doesn't allow octal values above 32 (0x20 in hex)
+                    if (HasOption(options, RegexOptions.ECMAScript) && currentVal >= 0x20)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            Debug.Assert(Position - start > 0);
```

Comment:
How do we know that? Other scan methods have affordance for finding no useful characters. Maybe add a comment (here or on method)?

Reply:
i'll add a comment.   ScanOctalCharacters should only be called when the caller knows that there is an octal token here.  so it should always make progress.

Reply:
#resolved